### PR TITLE
(PDB-1632) Transfer old AMQP messages at startup

### PR DESCRIPTION
**Pull request by rbrw**
At startup, transfer any messages found in the old
com.puppetlabs.puppetdb.commands queue to puppetlabs.puppetdb.commands
so that pending commands won't be ignored after an upgrade.

### CODE REVIEW

**Comment by ajroetker** [comment]
Will this require some dashboard changes as well?

**Comment by wkalt** [comment]
I think @ajroetker is right about that. Looks like the test failures may be caused by querying for inaccessible metrics:
https://github.com/puppetlabs/puppetdb/blob/master/acceptance/helper.rb#L562-562

**Comment by wkalt** [comment]
this doesn't seem to solve the root issue yet -- still can't see messages getting transferred on upgrade

**Comment by ajroetker** [comment]
We should also probably have some kind of test for this behavior as part of this PR

**Comment by rbrw** [comment]
Reworked following @senior's suggestion to try JMX commands.

File: src/puppetlabs/puppetdb/mq.clj [hunk 1]
```diff
@@ -264,6 +266,42 @@
...
+   destination))
+
+(defn-validated remove-local-queue!
+  "Deletes the named local queue.  Throws
+  {:type ::queue-not-found :name name} when the queue doesn't exist."
+  [name :- s/Str]
+  ;; When the queue doesn't exist, removeQueue provides no indication.
+  (jmx/invoke-signature local-broker-mb
+                        :removeQueue ["java.lang.String"] name))
+
```
Comment (rbrw) [inline src/puppetlabs/puppetdb/mq.clj]: Do we want any unit tests for these right now?  Related: https://tickets.puppetlabs.com/browse/PDB-1634

**Comment by wkalt** [comment]
+1, verified it works

**Comment by rbrw** [comment]
Now with (at least some) tests!

File: test/puppetlabs/puppetdb/mq_test.clj [hunk 2]
```diff
@@ -109,8 +110,34 @@
...
                            #"status 404"
                            (svc-utils/current-queue-depth))))))
+
+(deftest queue-operations
+  (with-test-broker "test" conn
+    (remove-queue! "test" "foo")
+    (is (thrown+? [:type ::mq/queue-not-found] (queue-size "test" "foo")))
+    (send-message! conn "foo" "1")
+    (send-message! conn "bar" "1")
+    (bounded-drain-into-vec! conn "foo" 1)
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/mq_test.clj]: is the idea here to test that bounded-drain-into-vec successfully empties the queue? If so do we need another check above this to ensure that the message made it in to start with?

Comment (rbrw) [inline test/puppetlabs/puppetdb/mq_test.clj]: No, this was just done so that we know that the message reached the queue (i.e. so we don't have to poll or similar).

Comment (wkalt) [inline test/puppetlabs/puppetdb/mq_test.clj]: makes sense to me now

**Comment by wkalt** [comment]
looks good to me

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 3]
```diff
@@ -222,6 +222,21 @@
   [max-frame-size url]
   (format "%s&wireFormat.maxFrameSize=%s&marshal=true" url max-frame-size))
 
+(defn- transfer-old-messages! [connection]
+  (let [[pending exists?]
```
Comment (senior) [inline src/puppetlabs/puppetdb/cli/services.clj]: Looks like we might be able to use [this](http://activemq.apache.org/maven/5.10.0/apidocs/org/apache/activemq/broker/jmx/BrokerViewMBean.html#getQueues()) JMX call to get a list of queues that the broker knows about. Then we wouldn't need to infer existence based on queue size (and handle the errors associated with that etc).

File: test/puppetlabs/puppetdb/mq_test.clj [hunk 4]
```diff
@@ -109,8 +110,34 @@
...
 
 (deftest test-jmx-enabled
-  (svc-utils/without-jmx
+  (without-jmx
    (svc-utils/with-puppetdb-instance
      (is (thrown-with-msg? clojure.lang.ExceptionInfo
                            #"status 404"
                            (svc-utils/current-queue-depth))))))
+
+(deftest queue-operations
```
Comment (senior) [inline test/puppetlabs/puppetdb/mq_test.clj]: I think this test should be broken up into separate tests for the various operations. With all the operations being tested at once, I find it hard to follow and separate what's being tested in my head.

File: src/puppetlabs/puppetdb/mq.clj [hunk 5]
```diff
@@ -264,6 +266,45 @@
...
+                        :moveMatchingMessagesTo
+                        ["java.lang.String" "java.lang.String"]
+                        "JMSMessageID is NOT NULL"
+                        destination))
+
+(defn-validated remove-queue!
+  "Deletes the named broker queue.  Does nothing if the queue doesn't exist."
+  [broker :- s/Str
+   queue :- s/Str]
+  ;; When the queue doesn't exist, removeQueue provides no indication.
```
Comment (senior) [inline src/puppetlabs/puppetdb/mq.clj]: Is this comment a restatement of the docstring?

File: src/puppetlabs/puppetdb/cli/services.clj (hunk 3 above)
Comment (senior) [inline src/puppetlabs/puppetdb/cli/services.clj]: Looking at this again, I don't know that I'd change it to the existence call. I can see wanting to log the messages there before the messages have been moved.

File: src/puppetlabs/puppetdb/mq.clj (hunk 5 above)
Comment (rbrw) [inline src/puppetlabs/puppetdb/mq.clj]: Practically, I suppose that's correct, though I'd left it because I had to figure it out by testing, and thought it might save someone else time (i.e. so they don't think it's possible, but we just chose to ignore the return value or something).  In any case, I'm happy to remove it.

File: test/puppetlabs/puppetdb/mq_test.clj (hunk 4 above)
Comment (rbrw) [inline test/puppetlabs/puppetdb/mq_test.clj]: Crossed my mind too.  I'll fix it.

**Comment by wkalt** [comment]
@pljenkinsro retest this please
